File: backend/app/schemas/cortage.py
import uuid
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class CortageBase(BaseModel):
    uniform_size: str | None = Field(default=None, max_length=20)
    fitting_status: Literal["pending", "fitting", "done"] = "pending"
    payment_status: Literal["belum_bayar", "dp", "lunas"] = "belum_bayar"
    price: int = Field(default=0, ge=0)
    notes: str | None = None


class CortageUpdate(BaseModel):
    uniform_size: str | None = Field(default=None, max_length=20)
    fitting_status: Literal["pending", "fitting", "done"] | None = None
    payment_status: Literal["belum_bayar", "dp", "lunas"] | None = None
    price: int | None = Field(default=None, ge=0)
    notes: str | None = None


class CortageResponse(CortageBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    wedding_id: uuid.UUID
    guest_id: uuid.UUID
    guest_name: str
    guest_phone: str | None = None
    guest_side: str | None = None
    guest_category: str | None = None
    created_at: datetime
    updated_at: datetime


# backwards compat aliases
BridesmaidBase = CortageBase
BridesmaidUpdate = CortageUpdate
BridesmaidResponse = CortageResponse

File: frontend/app/stores/cortage.ts
import { defineStore } from 'pinia'

export interface Cortage {
  id: string
  wedding_id: string
  guest_id: string
  guest_name: string
  guest_phone: string | null
  guest_side: string | null
  guest_category: string | null
  uniform_size: string | null
  fitting_status: 'pending' | 'fitting' | 'done'
  payment_status: 'belum_bayar' | 'dp' | 'lunas'
  price: number
  notes: string | null
  created_at: string
  updated_at: string
}

// backwards compat alias
export type Bridesmaid = Cortage

export const useCortageStore = defineStore('cortage', () => {
  const items = ref<Cortage[]>([])
  const loading = ref(false)
  const error = ref<string | null>(null)
  const weddingStore = useWeddingStore()
  const weddingId = computed(() => weddingStore.wedding?.id)

  const totalPrice = computed(() => items.value.reduce((s, i) => s + (i.price || 0), 0))
  const lunasCount = computed(() => items.value.filter((i) => i.payment_status === 'lunas').length)
  const fittingDoneCount = computed(() => items.value.filter((i) => i.fitting_status === 'done').length)

  async function fetchCortage() {
    const api = useApi()
    if (!weddingId.value) return
    loading.value = true
    error.value = null
    try {
      const res = await api<Cortage[]>(`/api/v1/weddings/${weddingId.value}/cortage`)
      items.value = res as unknown as Cortage[]
    } catch (err: unknown) {
      error.value = extractError(err)
    } finally {
      loading.value = false
    }
  }

  // alias for backwards compat
  const fetchBridesmaids = fetchCortage

  async function updateCortage(id: string, data: Partial<Cortage>) {
    const api = useApi()
    if (!weddingId.value) return
    const idx = items.value.findIndex((i) => i.id === id)
    const prev = idx !== -1 ? { ...items.value[idx] } : null
    if (idx !== -1) Object.assign(items.value[idx], data)
    try {
      const real = await api<Cortage>(`/api/v1/weddings/${weddingId.value}/cortage/${id}`, {
        method: 'PATCH',
        body: data,
      })
      if (idx !== -1) items.value[idx] = real as unknown as Cortage
      return real
    } catch (err) {
      if (prev && idx !== -1) items.value[idx] = prev as Cortage
      error.value = extractError(err)
      throw err
    }
  }

  const updateBridesmaid = updateCortage

  function extractError(err: unknown): string {
    const e = err as { data?: { detail?: unknown } }
    const d = e?.data?.detail
    if (typeof d === 'string') return d
    if (d && typeof d === 'object' && 'message' in d) return String((d as Record<string, unknown>).message)
    return 'Terjadi kesalahan.'
  }

  return { items, loading, error, totalPrice, lunasCount, fittingDoneCount, fetchCortage, fetchBridesmaids, updateCortage, updateBridesmaid }
})

// backwards compat
export const useBridesmaidStore = useCortageStore
